lib/advSearchOk.ml

```diff
@@ -34,22 +34,39 @@ let rec skip_no_spaces x i =
   else if String.unsafe_get x i != ' ' then skip_no_spaces x (i + 1)
   else i
 
-let string_incl x y =
-  let rec loop j_ini =
-    if j_ini = String.length y then false
-    else
-      let rec loop1 i j =
-        if i = String.length x then
-          if j = String.length y then true
-          else String.unsafe_get y j = ' ' || String.unsafe_get y (j - 1) = ' '
-        else if
-          j < String.length y && String.unsafe_get x i = String.unsafe_get y j
-        then loop1 (i + 1) (j + 1)
-        else loop (skip_spaces y (skip_no_spaces y j_ini))
-      in
-      loop1 0 j_ini
-  in
-  loop 0
+let string_incl =
+  let memo : (string, bool Util.StrMap.t) Hashtbl.t = Hashtbl.create 10 in
```

Comment:
What about using a map associating boolean values to keys of type `string * string` instead?